n_plus_one_example/src/main.rs

```diff
@@ -0,0 +1,25 @@
+pub mod db;
+use color_eyre::eyre::Result;
+use db::{connect_db, fetch_comments_by_post, fetch_post_ids};
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let (client, connection_handle) = connect_db().await?;
+
+    tokio::spawn(async move {
+        if let Err(e) = connection_handle.await {
+            eprintln!("Connection error: {}", e);
+        }
+    });
+
+    let post_ids = fetch_post_ids(&client).await?;
+    println!("Fetched {} posts.", post_ids.len());
+
+    // N + 1 Bug: For each post, fetch comments individually
+    for post_id in post_ids {
+        let comments = fetch_comments_by_post(&client, post_id).await?;
+        println!("Post ID {} has {} comments.", post_id, comments.len());
+    }
```

Comment:
Ya this is a perfect N+1!

To make it feel more 'real' maybe we can move this logic to a function called something like `comment_count_per_post` or something and have it return a HashMap of post_id to comment count

And if we wanted to make it _slightly_ more realistic, we could have the `fetch_posts` method return a `Post` struct and impl the `fetch_comments_by_post` on that struct.
The only reason to do that is cause thats how you often end up with N+1 bugs. You make nice structs (or models in other languages) for yourself, and give them nice methods like `fetch_comments` and then its really tempting to use them in a loop just cause they are there!

But the basic idea of this N+1 is great! Just some suggestions to tune it up for the 'game' lol

Reply:
Yo this literally insane! I need to take notes 🎉